From 2bddfb0875fc952b9d81cafee925a8c880a4f171 Mon Sep 17 00:00:00 2001
From: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
Date: Mon, 28 Jul 2025 18:00:33 +0530
Subject: [PATCH] UPSTREAM: DP: wifi: ath12k: Add API for ath12k_dp_peer search
 using address and ieee80211_sta

In the following scenarios, hostapd might not send NL command for delete
station to driver:

1. In congested environment with multiple clients, de-authentication
   frames from client might be missed in AP Rx

2. In roaming scenario, client can roam across the link and it is
   not mandatory for client to send the deauthentication frame to AP

3. Sometimes client may go to out of coverage and try to connect back
   with another radio with same mac address

In such cases, when NL command for add station is received in ath12k
driver without delete station command being received, ath12k_dp_peer
creation fails since there is an existing entry in driver
with the same mac address and this results in station connection failure.

This issue is introduced due to introduction of ath12k_dp_peer as part
of Unified Driver framework.

In order to address this, search for ath12k_dp_peer using mac address and
ieee80211_sta in creation and deletion paths which makes it unique.

Fixes: 70b5a1a98 (wifi: ath12k: Modularization of peer object Part-2)

Patch Dependency:
0116-0001-wifi-ath12k-Define-ath12k_dp_peer-structure-APIs-for.patch
0116-0002-wifi-ath12k-Attach-and-detach-ath12k_dp_link_peer-to.patch

Signed-off-by: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/debugfs_sta.c |  2 +-
 drivers/net/wireless/ath/ath12k/dp_cmn.h      |  2 +-
 drivers/net/wireless/ath/ath12k/dp_peer.c     | 33 +++++++++++++------
 drivers/net/wireless/ath/ath12k/dp_peer.h     |  4 ++-
 drivers/net/wireless/ath/ath12k/mac.c         |  6 ++--
 drivers/net/wireless/ath/ath12k/peer.c        |  2 +-
 6 files changed, 32 insertions(+), 17 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/debugfs_sta.c b/drivers/net/wireless/ath/ath12k/debugfs_sta.c
index 08b9845..71fccff 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs_sta.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs_sta.c
@@ -334,7 +334,7 @@ ath12k_dbg_sta_read_qos_msduq(struct file *file, char __user *user_buf,
 	wiphy_lock(ath12k_ar_to_hw(ar)->wiphy);
 	spin_lock_bh(&ah->dp_hw.peer_lock);
 
-	peer = ath12k_dp_peer_find(&ah->dp_hw, arsta->addr);
+	peer = ath12k_dp_peer_find_by_addr_and_sta(&ah->dp_hw, arsta->addr, sta);
 	if (!peer) {
 		goto ret;
 	}
diff --git a/drivers/net/wireless/ath/ath12k/dp_cmn.h b/drivers/net/wireless/ath/ath12k/dp_cmn.h
index 3cd4573..a684b43 100644
--- a/drivers/net/wireless/ath/ath12k/dp_cmn.h
+++ b/drivers/net/wireless/ath/ath12k/dp_cmn.h
@@ -103,7 +103,7 @@ void ath12k_dp_cmn_update_hw_links(struct ath12k_dp *dp,
 				   struct ath12k_hw_group *ag,
 				   struct ath12k *ar);
 int ath12k_dp_link_peer_assign(struct ath12k *ar, u8 vdev_id,
-			       u8 *dp_peer_addr, u8 *addr, u8 link_id,
+			       struct ieee80211_sta *sta, u8 *addr, u8 link_id,
 			       u32 hw_link_id, struct ieee80211_vif *vif);
 void ath12k_dp_link_peer_unassign(struct ath12k *ar, u8 vdev_id, u8 *addr);
 void ath12k_link_peer_get_sta_rate_info_stats(struct ath12k_dp *dp, const u8 *addr,
diff --git a/drivers/net/wireless/ath/ath12k/dp_peer.c b/drivers/net/wireless/ath/ath12k/dp_peer.c
index 3898c9b..8cd7e63 100644
--- a/drivers/net/wireless/ath/ath12k/dp_peer.c
+++ b/drivers/net/wireless/ath/ath12k/dp_peer.c
@@ -464,6 +464,21 @@ struct ath12k_dp_peer *ath12k_dp_peer_find(struct ath12k_dp_hw *dp_hw, u8 *addr)
 	return NULL;
 }
 
+struct ath12k_dp_peer *ath12k_dp_peer_find_by_addr_and_sta(struct ath12k_dp_hw *dp_hw, u8 *addr,
+							   struct ieee80211_sta *sta)
+{
+	struct ath12k_dp_peer *dp_peer;
+
+	lockdep_assert_held(&dp_hw->peer_lock);
+
+	list_for_each_entry(dp_peer, &dp_hw->peers, list) {
+		if (ether_addr_equal(dp_peer->addr, addr) && (dp_peer->sta == sta))
+			return dp_peer;
+	}
+
+	return NULL;
+}
+
 #define PEER_TABLE_SOC_ID_SHIFT        10
 
 u16 ath12k_dp_peer_get_peerid_index(struct ath12k_dp *dp, u16 peer_id)
@@ -520,7 +535,7 @@ int ath12k_dp_peer_create(struct ath12k_dp_hw *dp_hw, u8 *addr,
 	struct wireless_dev *wdev;
 
 	spin_lock_bh(&dp_hw->peer_lock);
-	dp_peer = ath12k_dp_peer_find(dp_hw, addr);
+	dp_peer = ath12k_dp_peer_find_by_addr_and_sta(dp_hw, addr, params->sta);
 	spin_unlock_bh(&dp_hw->peer_lock);
 
 	if (dp_peer)
@@ -556,14 +571,14 @@ int ath12k_dp_peer_create(struct ath12k_dp_hw *dp_hw, u8 *addr,
 	return 0;
 }
 
-void ath12k_dp_peer_delete(struct ath12k_dp_hw *dp_hw, u8 *addr)
+void ath12k_dp_peer_delete(struct ath12k_dp_hw *dp_hw, u8 *addr, struct ieee80211_sta *sta)
 {
 	struct ath12k_dp_peer *dp_peer;
 	u16 peerid_index;
 
 	spin_lock_bh(&dp_hw->peer_lock);
 
-	dp_peer = ath12k_dp_peer_find(dp_hw, addr);
+	dp_peer = ath12k_dp_peer_find_by_addr_and_sta(dp_hw, addr, sta);
 	if (!dp_peer) {
 		spin_unlock_bh(&dp_hw->peer_lock);
 		return;
@@ -584,7 +599,7 @@ void ath12k_dp_peer_delete(struct ath12k_dp_hw *dp_hw, u8 *addr)
 }
 
 int ath12k_dp_link_peer_assign(struct ath12k *ar, u8 vdev_id,
-			       u8 *dp_peer_addr, u8 *addr, u8 link_id,
+			       struct ieee80211_sta *sta, u8 *addr, u8 link_id,
 			       u32 hw_link_id, struct ieee80211_vif *vif)
 {
 	struct ath12k_pdev_dp *dp_pdev = &ar->dp;
@@ -594,16 +609,14 @@ int ath12k_dp_link_peer_assign(struct ath12k *ar, u8 vdev_id,
 	struct ath12k_dp_link_peer *peer;
 	u16 peerid_index;
 	int ret;
-	u8 *dp_peer_mac = dp_peer_addr;
+	u8 *dp_peer_mac = !sta ? addr : sta->addr;
 
-	if (!dp_peer_addr) {
+	if (!sta) {
 		struct ath12k_dp_peer_create_params params = {0};
 
 		params.is_vdev_peer = true;
 
 		ath12k_dp_peer_create(dp_hw, addr, &params, vif);
-
-		dp_peer_mac = addr;
 	}
 
 	spin_lock_bh(&dp->dp_lock);
@@ -616,7 +629,7 @@ int ath12k_dp_link_peer_assign(struct ath12k *ar, u8 vdev_id,
 
 	spin_lock_bh(&dp_hw->peer_lock);
 
-	dp_peer = ath12k_dp_peer_find(dp_hw, dp_peer_mac);
+	dp_peer = ath12k_dp_peer_find_by_addr_and_sta(dp_hw, dp_peer_mac, sta);
 	if (!dp_peer) {
 		ret = -ENOENT;
 		goto err_dp_peer;
@@ -712,7 +725,7 @@ void ath12k_dp_link_peer_unassign(struct ath12k *ar, u8 vdev_id, u8 *addr)
 	synchronize_rcu();
 
 	if (dp_peer->is_vdev_peer)
-		ath12k_dp_peer_delete(dp_hw, addr);
+		ath12k_dp_peer_delete(dp_hw, addr, NULL);
 }
 
 void ath12k_link_peer_get_sta_rate_info_stats(struct ath12k_dp *dp, const u8 *addr,
diff --git a/drivers/net/wireless/ath/ath12k/dp_peer.h b/drivers/net/wireless/ath/ath12k/dp_peer.h
index 15b2309..eb6280f 100644
--- a/drivers/net/wireless/ath/ath12k/dp_peer.h
+++ b/drivers/net/wireless/ath/ath12k/dp_peer.h
@@ -217,6 +217,8 @@ void ath12k_peer_map_event(struct ath12k_base *ab, u8 vdev_id, u16 peer_id,
 			   u8 *mac_addr, u16 ast_hash, u16 hw_peer_id);
 struct ath12k_dp_peer *ath12k_dp_peer_find(struct ath12k_dp_hw *dp_hw,
 					   u8 *addr);
+struct ath12k_dp_peer *ath12k_dp_peer_find_by_addr_and_sta(struct ath12k_dp_hw *dp_hw,
+							   u8 *addr, struct ieee80211_sta *sta);
 struct ath12k_dp_link_peer *
 ath12k_dp_link_peer_find_by_vdev_id_and_addr(struct ath12k_dp *dp,
 					     int vdev_id, const u8 *addr);
@@ -237,7 +239,7 @@ int ath12k_dp_link_peer_rhash_delete(struct ath12k_dp *dp,
 int ath12k_dp_peer_create(struct ath12k_dp_hw *dp_hw, u8 *addr,
 			  struct ath12k_dp_peer_create_params *params,
 			  struct ieee80211_vif *vif);
-void ath12k_dp_peer_delete(struct ath12k_dp_hw *dp_hw, u8 *addr);
+void ath12k_dp_peer_delete(struct ath12k_dp_hw *dp_hw, u8 *addr, struct ieee80211_sta *sta);
 struct ath12k_dp_peer *ath12k_dp_peer_find_by_peerid_index(struct ath12k_dp *dp,
 							   struct ath12k_pdev_dp *dp_pdev,
 							   u16 peer_id);
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 53e24a7..0c18365 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1426,7 +1426,7 @@ void ath12k_mac_peer_cleanup_all(struct ath12k *ar)
 	/* Delete all the self dp_peers on asserted radio
 	 */
 	list_for_each_entry_safe_reverse(arvif, tmp_vif, &ar->arvifs, list) {
-		ath12k_dp_peer_delete(dp_hw, arvif->bssid);
+		ath12k_dp_peer_delete(dp_hw, arvif->bssid, NULL);
 		arvif->num_stations = 0;
 	}
 
@@ -11229,7 +11229,7 @@ ml_station_remove:
 
 	if (old_state == IEEE80211_STA_NONE &&
 	    new_state == IEEE80211_STA_NOTEXIST) {
-		ath12k_dp_peer_delete(&ah->dp_hw, sta->addr);
+		ath12k_dp_peer_delete(&ah->dp_hw, sta->addr, sta);
 		wiphy_work_cancel(hw->wiphy, &ahsta->set_4addr_wk);
 	}
 
@@ -11242,7 +11242,7 @@ ml_station_remove:
 
 peer_delete:
 	if (ret)
-		ath12k_dp_peer_delete(&ah->dp_hw, sta->addr);
+		ath12k_dp_peer_delete(&ah->dp_hw, sta->addr, sta);
 ml_peer_id_free:
 	if (ret)
 		ath12k_peer_ml_free(ah, ahsta);
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index e7c8b0e..1e56e64 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -263,7 +263,7 @@ int ath12k_peer_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	spin_unlock_bh(&ar->ab->dp->dp_lock);
 
 	ath12k_dp_link_peer_assign(ar, arvif->vdev_id,
-				   sta ? sta->addr : NULL, arg->peer_addr,
+				   sta, arg->peer_addr,
 				   link_id, ar->hw_link_id, vif);
 
 	if (vif->type == NL80211_IFTYPE_AP)
-- 
2.34.1

